src/main/kotlin/us/ihmc/cd/RemoteExtension.kt

```diff
@@ -0,0 +1,136 @@
+   {
+      session(address, {sshClient -> sshClient.authPassword(username, password)} , action)
+   }
+
+   fun session(address: String, username: String, action: Action<RemoteConnection>)
+   {
+      session(address, {sshClient -> authWithSSHKey(username, sshClient)} , action)
+   }
+
+   /**
+    * Replicate OpenSSH functionality where users can name private keys whatever they want.
+    */
+   private fun authWithSSHKey(username: String, sshClient: SSHClient)
+   {
+      val sshDir = OpenSSHKnownHosts.detectSSHDir()
+
+      if (sshDir.isDirectory)
+      {
+         val list = Files.list(sshDir.toPath())
+
+         val privateKeyFiles = arrayListOf<String>()
+         for (path in list)
+         {
+            if (Files.isRegularFile(path)
+               && path.fileName.toString() != "config"
+               && path.fileName.toString() != "known_hosts"
+               && !path.fileName.toString().endsWith(".pub"))
+            {
+               val absoluteNormalizedString = path.toAbsolutePath().normalize().toString()
+               privateKeyFiles.add(absoluteNormalizedString)
+            }
+         }
+
+         LogTools.info("Passing keys to authPublicKey: $privateKeyFiles")
+
+         sshClient.authPublickey(username, *privateKeyFiles.toTypedArray())
+      }
+   }
+
+   class RemoteConnection(val ssh: SSHClient, val sftp: SFTPClient)
+   {
+      fun exec(command: String, timeout: Double = 5.0)
+      {
+         var session: Session? = null
+         try
+         {
+            session = ssh.startSession()
+
+            LogTools.quiet("Executing on ${ssh.remoteHostname}: \"$command\"")
+            val sshjCommand = session.exec(command)
+            LogTools.quiet(IOUtils.readFully(sshjCommand.inputStream).toString())
+            sshjCommand.join((timeout * 1e9).toLong(), TimeUnit.NANOSECONDS)
+            LogTools.quiet("** exit status: " + sshjCommand.exitStatus)
+         }
+         finally
+         {
+            try
+            {
+               session?.close()
+            }
+            catch (e: IOException)
+            {
+               // do nothing
+            }
+         }
+      }
+
+      fun put(source: String, dest: String)
+      {
+         LogTools.quiet("Putting $source to ${ssh.remoteHostname}:$dest")
+         sftp.put(source, dest)
+      }
+
+      fun get(source: String, dest: String)
+      {
+         LogTools.quiet("Getting ${ssh.remoteHostname}:$source to $dest")
+         sftp.get(source, dest)
+      }
+   }
+
+   fun session(address: String, authenticate: (SSHClient) -> Unit, action: Action<RemoteConnection>)
+   {
+      val sshClient = SSHClient()
+
+      val sshDir = OpenSSHKnownHosts.detectSSHDir()
+
+      if (sshDir.resolve("known_hosts").isFile)
+      {
+         sshClient.loadKnownHosts()
+      }
+      else
+      {
+         LogTools.warn("Could not find known_hosts file. Disabling host key verification.")
```

Comment:
I think this should require a `-PignoreHostIdentity=true` or something like that instead of immediately disabling security.

Reply:
I don't think this is a huge issue. Can you argue against it? Or, can I just keep this?

Reply:
There's no point in doing any of the other stuff if ignoring it is an option, that's silly. And now that we know about the security vulnerability, we should fix it

Reply:
What do you propose to fix it? I had the idea of trying to intelligently find a `known_hosts` on Windows by checking a bunch of relevant known locations.

Reply:
You could do that, but I was just saying to have an "if" statement that skips the checks all together if a gradle property is there

Reply:
Right, but that isn't a "fix". If we have the skip thing, I think we should also have some potential fix. Without any fix, it seems like we'd just be having most Windows users run the skip anyway, which is basically no different than my change.

Reply:
It could also be an actual programmatic parameter too